### PR TITLE
VZ-7635:Reject Invalid changes to opensearch CR

### PR DESCRIPTION
**Pull request by sk2701**
There are 2 criteria for determining whether a change to opensearch CR is valid or not
1. The number of master nodes, data nodes and ingest nodes should be at least 3,2 and 1 respectively, if the cluster is not single node.
2. If we are scaling down the number of master and data nodes, the scaling down shouldn't be by more than half.
Therefore,minimum number of master nodes are max(3,(current master nodes)/2) and min number of data nodes are max(2,(current data nodes)/2)

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 1]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
 	if err := common.CompareStorageOverridesV1Beta1(old, new, ComponentJSONName); err != nil {
 		return err
 	}
+	if old.Spec.Components.OpenSearch == nil {
```
Comment (adiforluls) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:207]: Should the webhook logic be in both ValidateInstall and ValidateUpdate? (also ValidateInstallV1Beta1 and ValidateUpdateV1Beta1) I think yes.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 2]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+		return nil
+	}
+	opensearchOld := old.Spec.Components.OpenSearch
+	if new.Spec.Components.OpenSearch == nil {
+		return nil
```
Comment (adiforluls) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:212]: Improve error handling. If your func is supposed to return an error (among other things), then your logic should be something like "return from the func midway if there's an error, return nil at the end (i.e. success)".

Comment (sk2701) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:212]: okay

File: platform-operator/controllers/verrazzano/component/opensearch/validation.go [hunk 3]
```diff
@@ -48,5 +50,35 @@ func validateNoDuplicateNodeGroups(opensearch *v1beta1.OpenSearchComponent) erro
...
+		for _, role := range group.Roles {
+			numberNodes[role] += group.Replicas
+			sum += group.Replicas
+		}
+	}
+	if sum <= 1 {
+		return numberNodes, nil
+	}
+
+	if numberNodes["master"] < 3 {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/validation.go:72]: instead of string constants use the vmov1.NodeRole values

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 4]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+	}
+	opensearchNew := new.Spec.Components.OpenSearch
+
+	numNodesold, _ := GetNodesNumber(opensearchOld)
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:216]: why is it ignoring the error here?

File: platform-operator/controllers/verrazzano/component/opensearch/validation.go [hunk 5]
```diff
@@ -48,5 +50,35 @@ func validateNoDuplicateNodeGroups(opensearch *v1beta1.OpenSearchComponent) erro
 			return fmt.Errorf("OpenSearch node group name is duplicated or invalid: %v", err)
 		}
 	}
+	if _, err := GetNodesNumber(opensearch); err != nil {
+		return err
+	}
 	return nil
 }
+
+func GetNodesNumber(opensearch *v1beta1.OpenSearchComponent) (map[vmov1.NodeRole]int32, error) {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/validation.go:59]: suggest to reorganize this a bit:
- GetNodesNumber (can you change this name to `GetNodeRoleCounts`
- Can you move the validation to another function? It would be great if this function just returned the role counts and validation was handled separately

File: platform-operator/controllers/verrazzano/component/opensearch/validation.go [hunk 6]
```diff
@@ -48,5 +50,35 @@ func validateNoDuplicateNodeGroups(opensearch *v1beta1.OpenSearchComponent) erro
...
+	numberNodes := make(map[vmov1.NodeRole]int32)
+	sum := int32(0)
+	for _, group := range opensearch.Nodes {
+		for _, role := range group.Roles {
+			numberNodes[role] += group.Replicas
+			sum += group.Replicas
+		}
+	}
+	if sum <= 1 {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/validation.go:68]: I valid single node cluster may have 1 or 3 roles. It matters that they are on the same _node_ though. E.g., if I have one node with master role, or one node with master,data,ingest. Either of those configurations is a valid single node cluster.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 7]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+	numNodesnew, _ := GetNodesNumber(opensearchNew)
+	sum := numNodesnew["master"] + numNodesnew["data"] + numNodesnew["ingest"]
+	for role, replicas := range numNodesold {
+
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:220]: can you remove the extra whitespace

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go (hunk 2 above)
Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:212]: Parameter validation should be done at the beginning of a function

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go (hunk 4 above)
Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:216]: and on the next line

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 8]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+		if role != "ingest" && replicas-numNodesnew[role] > numNodesnew[role] && sum > 1 {
```
[the remark this answers is not in the document]
Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:221]: I also don't understand this logic. What is this trying to say? I also don't see where `replicas-numNodesnew` is coming from

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 9]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+			return errors.New("The number of nodes can't be scaled by more than half")
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:222]: What if user needs to scale down anyway? You might need to do something outside the normal parameters in exceptional scenarios.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 10]
```diff
@@ -203,6 +204,26 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+
+		}
+
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:225]: remove the extra whitespace

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go [hunk 11]
```diff
@@ -1096,3 +1097,99 @@ func TestValidateUpdate(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateInstallCR(t *testing.T) {
+	trueVal := true
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go:1102]: Use parameterized tests if you are testing multiple scenarios in the same spec.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go (hunk 9 above)
Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:222]: We should discuss on monday, I am not sure how to handle this. It's valid to want to scale down your cluster in such a way that make break the underlying nodes (in some circumstances).

Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:222]: If you are unable to scale down or otherwise modify the cluster topology when you need to, it's very frustrating.

Comment (sk2701) [reply platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:222]: okay

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 12]
```diff
@@ -203,6 +205,19 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
 	}
+	if old.Spec.Components.OpenSearch != nil && new.Spec.Components.OpenSearch != nil {
+		opensearchOld := old.Spec.Components.OpenSearch
+		opensearchNew := new.Spec.Components.OpenSearch
+
+		numNodesold, _ := GetNodeRoleCounts(opensearchOld)
+		numNodesnew, totalNodesNew := GetNodeRoleCounts(opensearchNew)
+
+		for role, replicas := range numNodesold {
+			if role != vmov1.NodeRole("ingest") && replicas > 2*numNodesnew[role] && totalNodesNew > int32(1) {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:216]: Please go through and remove any string constants like "ingest", "data", or "master", and use the vmo noderole type alias instead. In this case, it'd look like:
```if role != vmov1.IngestRole```

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 13]
```diff
@@ -203,6 +205,19 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+				return errors.New("The number of nodes can't be scaled by more than half at once")
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:217]: Error should describe the type of node that is being scaled

File: platform-operator/controllers/verrazzano/component/opensearch/validation.go [hunk 14]
```diff
@@ -48,5 +50,30 @@ func validateNoDuplicateNodeGroups(opensearch *v1beta1.OpenSearchComponent) erro
...
 		}
 	}
+	numberNodes, totalNode := GetNodeRoleCounts(opensearch)
+
+	if totalNode > int32(1) {
+		if numberNodes[vmov1.NodeRole("master")] < 3 {
+			return errors.New("Number of master nodes should be atleast 3")
+		}
+		if numberNodes[vmov1.NodeRole("data")] < 2 {
+			return errors.New("Number of data nodes should be atleast 2")
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/validation.go:60]: atleast -> at least

File: platform-operator/controllers/verrazzano/component/opensearch/validation.go [hunk 15]
```diff
@@ -48,5 +50,30 @@ func validateNoDuplicateNodeGroups(opensearch *v1beta1.OpenSearchComponent) erro
...
+		}
+		if numberNodes[vmov1.NodeRole("ingest")] < 1 {
+			return errors.New("Number of ingest nodes should be atleast 1")
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/opensearch/validation.go:63]: atleast -> at least

**Review by anders-swanson — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 16]
```diff
@@ -203,6 +205,19 @@ func (o opensearchComponent) ValidateUpdateV1Beta1(old *installv1beta1.Verrazzan
...
+	if old.Spec.Components.OpenSearch != nil && new.Spec.Components.OpenSearch != nil {
+		opensearchOld := old.Spec.Components.OpenSearch
+		opensearchNew := new.Spec.Components.OpenSearch
+
+		numNodesold, _ := GetNodeRoleCounts(opensearchOld)
+		numNodesnew, totalNodesNew := GetNodeRoleCounts(opensearchNew)
+
+		for role, replicas := range numNodesold {
+			if role != vmov1.IngestRole && replicas > 2*numNodesnew[role] && totalNodesNew > int32(1) {
+				return fmt.Errorf("The number of %v nodes can't be scaled by more than half at once", role)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:217]: I believe this only applies to scaling them down.  The message should say that.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go [hunk 17]
```diff
@@ -1096,3 +1097,158 @@ func TestValidateUpdate(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateInstallCR(t *testing.T) {
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go:1101]: Needs standard test comment.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go [hunk 18]
```diff
@@ -1096,3 +1097,158 @@ func TestValidateUpdate(t *testing.T) {
...
+	trueVal := true
+	tests := []struct {
+		name    string
+		vzCR    *vzapi.Verrazzano
+		wantErr bool
+	}{
+		{
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go:1108]: Each of these should have a standard test comment describing the purpose of the test parameters.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go [hunk 19]
```diff
@@ -1096,3 +1097,158 @@ func TestValidateUpdate(t *testing.T) {
...
+
+func TestValidateUpdateCR(t *testing.T) {
+	trueVal := true
+	tests := []struct {
+		name    string
+		old     *vzapi.Verrazzano
+		new     *vzapi.Verrazzano
+		wantErr bool
+	}{
+		{
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component_test.go:1171]: Comments per set of test parameters.

File: platform-operator/controllers/verrazzano/component/opensearch/validation.go [hunk 20]
```diff
@@ -48,5 +50,30 @@ func validateNoDuplicateNodeGroups(opensearch *v1beta1.OpenSearchComponent) erro
...
+		if numberNodes[vmov1.DataRole] < 2 {
+			return errors.New("Number of data nodes should be at least 2")
+		}
+		if numberNodes[vmov1.IngestRole] < 1 {
+			return errors.New("Number of ingest nodes should be at least 1")
+		}
+	}
 	return nil
 }
+
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/opensearch/validation.go:68]: Does this need to be public?
Needs comments.

File: platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go [hunk 21]
```diff
@@ -4,13 +4,15 @@
 package opensearch
 
 import (
+	//"errors"
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/opensearch/opensearch_component.go:7]: Cleanup

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go [hunk 22]
```diff
@@ -56,13 +63,76 @@ func (v *RequirementsValidatorV1alpha1) Handle(ctx context.Context, req admissio
...
 }
+
+func getWarningArrayWithOS(vz *v1alpha1.Verrazzano) []string {
+	var warnings []string
+	vzv1beta1 := &v1beta1.Verrazzano{}
+	if err := vz.ConvertTo(vzv1beta1); err == nil {
+		if vzv1beta1.Spec.Components.OpenSearch != nil {
+			opensearch := vzv1beta1.Spec.Components.OpenSearch
+			numberNodes, totalNode := GetNodeRoleCounts(opensearch)
+			if totalNode > int32(1) {
```
Comment (anders-swanson) [inline platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go:121]: This code seems duplicated in both packages. Can you consolidate?

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go [hunk 23]
```diff
@@ -56,14 +62,78 @@ func (v *RequirementsValidatorV1beta1) Handle(ctx context.Context, req admission
...
+	}
+	if len(warnings) > 0 {
 		return admission.Allowed("").WithWarnings(warnings...)
 	}
+	return response
+}
+func validateUpdateForNodesv1beta1(log *zap.SugaredLogger, client client.Client, oldvz v1beta1.Verrazzano, newvz *v1beta1.Verrazzano) admission.Response {
+	response := admission.Allowed("")
+	warnings := getWarningArrayWithOSv1beta1(newvz)
+	if newvz.Spec.Components.OpenSearch != nil && oldvz.Spec.Components.OpenSearch != nil {
```
Comment (anders-swanson) [inline platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go:80]: Code is duplicated again, if using conversion you can make common function

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go [hunk 24]
```diff
@@ -56,14 +62,78 @@ func (v *RequirementsValidatorV1beta1) Handle(ctx context.Context, req admission
...
+		numNodesnew, totalNodesNew := GetNodeRoleCounts(opensearchNew)
+
+		for role, replicas := range numNodesold {
+			if role != vmov1.IngestRole && replicas > 2*numNodesnew[role] && totalNodesNew > int32(1) {
+				strRole := "The number of " + string(role) + " nodes shouldn't be scaled down by more than half at once"
+				warnings = append(warnings, strRole)
+			}
+		}
+	}
+	if errs := vzchecks.PrerequisiteCheck(client, vzchecks.ProfileType(newvz.Spec.Profile)); len(errs) > 0 {
```
Comment (anders-swanson) [inline platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go:94]: Are these warnings related to opensearch? If not, it shouldn't be in this function (or the function name should change)

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go [hunk 25]
```diff
@@ -56,13 +63,76 @@ func (v *RequirementsValidatorV1alpha1) Handle(ctx context.Context, req admissio
...
+
+			for role, replicas := range numNodesold {
+				if role != vmov1.IngestRole && replicas > 2*numNodesnew[role] && totalNodesNew > int32(1) {
+					strRole := "The number of " + string(role) + " nodes shouldn't be scaled down by more than half at once"
+					warnings = append(warnings, strRole)
+				}
+			}
+		}
+	}
+	if errs := vzchecks.PrerequisiteCheck(client, vzchecks.ProfileType(newvz.Spec.Profile)); len(errs) > 0 {
```
Comment (anders-swanson) [inline platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go:102]: I think you can simplify the code a little: 
```
for _, err := range vzchecks.PrerequisiteCheck(client, vzchecks.ProfileType(newvz.Spec.Profile)) {
  ...
}
```

could be wrong but give it a try

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go [hunk 26]
```diff
@@ -56,13 +63,76 @@ func (v *RequirementsValidatorV1alpha1) Handle(ctx context.Context, req admissio
...
+		if newvzv1beta1.Spec.Components.OpenSearch != nil && oldvzv1beta1.Spec.Components.OpenSearch != nil {
+			opensearchNew := newvzv1beta1.Spec.Components.OpenSearch
+			opensearchOld := oldvzv1beta1.Spec.Components.OpenSearch
+
+			numNodesold, _ := GetNodeRoleCounts(opensearchOld)
+			numNodesnew, totalNodesNew := GetNodeRoleCounts(opensearchNew)
+
+			for role, replicas := range numNodesold {
+				if role != vmov1.IngestRole && replicas > 2*numNodesnew[role] && totalNodesNew > int32(1) {
+					strRole := "The number of " + string(role) + " nodes shouldn't be scaled down by more than half at once"
```
Comment (anders-swanson) [inline platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go:96]: If this warning is coming from the admission webhook, do we want to link back to our documentation for an explanation?

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go (hunk 24 above)
Comment (sk2701) [reply platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go:94]: Changed the function name

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go (hunk 23 above)
Comment (sk2701) [reply platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1beta1.go:80]: reduced the corresponding part in v1alpha1 and calling this function there after conversion

File: platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go (hunk 26 above)
Comment (sk2701) [reply platform-operator/apis/verrazzano/webhooks/requirements_webhook_v1alpha1.go]: It's just a warning so I guess it's not needed. However if you want I can make the changes or the warning can be made more detailed

**Review by anders-swanson — APPROVED** [review]

**Review by kminder — APPROVED** [review]
Approved.

**Review by kminder — APPROVED** [review]
Approved.
